lib/libc/audio/lib_buffer.c

```diff
@@ -0,0 +1,208 @@
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: apb_semtake
+ *
+ *       Take an Audio Pipeline Buffer.
+ *
+ ****************************************************************************/
+
+static void apb_semtake(FAR struct ap_buffer_s *apb)
+{
+	/* Take the semaphore (perhaps waiting) */
+
+	while (sem_wait(&apb->sem) != 0) {
+		/* The only case that an error should occr here is if
+		 * the wait was awakened by a signal.
+		 */
+
+		ASSERT(errno == EINTR);
+	}
+}
+
+/****************************************************************************
+ * Name: apb_semgive
+ ****************************************************************************/
+
+#define apb_semgive(b) sem_post(&b->sem)
+
+/****************************************************************************
+ * Name: apb_alloc
+ *
+ * Allocate an Audio Pipeline Buffer for use in the Audio sub-system.  This
+ * will perform the actual allocate based on buffer data format, number of
+ * channels, etc. and prepare the buffer for consumption.
+ *
+ ****************************************************************************/
+
+int apb_alloc(FAR struct audio_buf_desc_s *bufdesc)
+{
+	uint32_t bufsize;
+	int ret;
+	struct ap_buffer_s *apb;
+
+	DEBUGASSERT(bufdesc->u.ppBuffer != NULL);
+
+	/* Perform a user mode allocation */
+
+	bufsize = sizeof(struct ap_buffer_s) + bufdesc->numbytes;
+	apb = lib_umalloc(bufsize);
+	*bufdesc->u.ppBuffer = apb;
+
+	/* Test if the allocation was successful or not */
+
+	if (*bufdesc->u.ppBuffer == NULL) {
+		ret = -ENOMEM;
+	} else {
+		/* Populate the buffer contents */
+
+		memset(apb, 0, bufsize);
+		apb->i.channels = 1;
+		apb->crefs = 1;
+		apb->nmaxbytes = bufdesc->numbytes;
+		apb->nbytes = 0;
+		apb->flags = 0;
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		apb->session = bufdesc->session;
+#endif
+
+		sem_init(&apb->sem, 0, 1);
+		ret = sizeof(struct audio_buf_desc_s);
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: apb_free
+ *
+ * Free's a previously allocated or referenced Audio Pipeline Buffer
+ *
+ ****************************************************************************/
+
+void apb_free(FAR struct ap_buffer_s *apb)
+{
+	int refcount;
+
+	/* Perform a reference count decrement and possibly release the memory */
+
+	apb_semtake(apb);
+	refcount = apb->crefs--;
+	apb_semgive(apb);
+
+	if (refcount <= 1) {
```

Comment:
Can't be done much. Coz, first it is a library and second it is void return function. Even if this code catches a bad scenario, it can't reply back anything. Hence as it is, better to leave the intelligence to application to track if the buffer is valid or not.